keel-sql/src/main/kotlin/com/netflix/spinnaker/keel/sql/SqlArtifactRepository.kt

```diff
@@ -215,6 +221,84 @@ class SqlArtifactRepository(
+          .select(
+            ENVIRONMENT_ARTIFACT_VERSIONS.ARTIFACT_VERSION,
+            ENVIRONMENT_ARTIFACT_VERSIONS.APPROVED_AT,
+            ENVIRONMENT_ARTIFACT_VERSIONS.DEPLOYED_AT
+          )
+          .from(
+            ENVIRONMENT_ARTIFACT_VERSIONS,
+            DELIVERY_ARTIFACT,
+            ENVIRONMENT,
+            DELIVERY_CONFIG
+          )
+          .where(DELIVERY_ARTIFACT.UID.eq(ENVIRONMENT_ARTIFACT_VERSIONS.ARTIFACT_UID))
+          .and(DELIVERY_ARTIFACT.NAME.eq(artifact.name))
+          .and(DELIVERY_ARTIFACT.TYPE.eq(artifact.type.name))
+          .and(ENVIRONMENT.UID.eq(ENVIRONMENT_ARTIFACT_VERSIONS.ENVIRONMENT_UID))
+          .and(ENVIRONMENT.DELIVERY_CONFIG_UID.eq(DELIVERY_CONFIG.UID))
+          .and(DELIVERY_CONFIG.NAME.eq(deliveryConfig.name))
+          .and(ENVIRONMENT.NAME.eq(environment.name))
+        val pendingVersions = jooq
+          .select(
+            DELIVERY_ARTIFACT_VERSION.VERSION,
+            castNull(LocalDateTime::class.java),
+            castNull(LocalDateTime::class.java)
+          )
+          .from(
+            DELIVERY_ARTIFACT_VERSION,
+            DELIVERY_ARTIFACT,
+            ENVIRONMENT,
+            DELIVERY_CONFIG
+          )
+          .where(DELIVERY_ARTIFACT.UID.eq(DELIVERY_ARTIFACT_VERSION.DELIVERY_ARTIFACT_UID))
+          .and(DELIVERY_ARTIFACT.NAME.eq(artifact.name))
+          .and(DELIVERY_ARTIFACT.TYPE.eq(artifact.type.name))
+          .and(ENVIRONMENT.DELIVERY_CONFIG_UID.eq(DELIVERY_CONFIG.UID))
+          .and(DELIVERY_CONFIG.NAME.eq(deliveryConfig.name))
+          .and(ENVIRONMENT.NAME.eq(environment.name))
+          .andNotExists(
+            selectOne()
+              .from(ENVIRONMENT_ARTIFACT_VERSIONS)
+              .where(ENVIRONMENT_ARTIFACT_VERSIONS.ARTIFACT_VERSION.eq(DELIVERY_ARTIFACT_VERSION.VERSION))
+              .and(ENVIRONMENT_ARTIFACT_VERSIONS.ENVIRONMENT_UID.eq(ENVIRONMENT.UID))
+          )
+        val versions = versionsInEnvironment
+          .unionAll(pendingVersions)
+          .fetch()
+          .map { (version, approvedAt, deployedAt) ->
+            ArtifactVersionResult(version, approvedAt, deployedAt)
+          }
+        val deployed = versions.filter { it.deployedAt != null }
+        val current = deployed.maxBy { checkNotNull(it.deployedAt) }
+        val previous = deployed.filterNot { it == current }
+        val deploying = versions.filter { it.approvedAt != null && it.deployedAt == null }.maxBy { checkNotNull(it.approvedAt) }
+        val pending = versions.filter { it.approvedAt == null }
+        ArtifactVersions(
+          name = artifact.name,
+          type = artifact.type,
+          versions = ArtifactVersionStatus(
+            current = current?.version,
+            deploying = deploying?.version,
+            pending = pending.map { it.version }.sortedWith(artifact.versioningStrategy.comparator),
+            previous = previous.map { it.version }.sortedWith(artifact.versioningStrategy.comparator)
```

Comment:
@emjburns good job with that `versioningStrategy`. I was worried sorting this stuff was going to be a PITA.

Reply:
Thanks!